src/constraint.rs

```diff
@@ -977,3 +993,83 @@ impl ConstraintValidator for CodepointLengthConstraint {
         Ok(())
     }
 }
+
+/// Implements a `element` constraint
+/// [element]: https://amzn.github.io/ion-schema/docs/spec.html#element
+#[derive(Debug, Clone, PartialEq)]
+pub struct ElementConstraint {
+    type_id: TypeId,
+}
+
+impl ElementConstraint {
+    pub fn new(type_id: TypeId) -> Self {
+        Self { type_id }
+    }
+
+    /// Tries to create an element constraint from the given OwnedElement
+    pub fn resolve_from_isl_constraint(
+        type_reference: &IslTypeRef,
+        type_store: &mut TypeStore,
+        pending_types: &mut PendingTypes,
+    ) -> IonSchemaResult<Self> {
+        let type_id =
+            IslTypeRef::resolve_type_reference(type_reference, type_store, pending_types)?;
+        Ok(ElementConstraint::new(type_id))
+    }
+}
+
+impl ConstraintValidator for ElementConstraint {
+    fn validate(&self, value: &OwnedElement, type_store: &TypeStore) -> ValidationResult {
+        let mut violations: Vec<Violation> = vec![];
+
+        // Check for null container
+        if value.is_null() {
+            return Err(Violation::new(
+                "element",
+                ViolationCode::TypeMismatched,
+                &format!("expected a container found {:?}", value),
+            ));
+        }
+
+        let type_def = type_store.get_type_by_id(self.type_id).unwrap();
```

Comment:
I think this `unwrap` is safe because we've already validated the type ID. If that's right, could you add a comment noting that?